File: api/generate.py
"""Vercel Serverless Function - 汉字合体字生成"""

import json
import os

from http.server import BaseHTTPRequestHandler
from urllib.request import Request, urlopen
from urllib.error import HTTPError, URLError
from pathlib import Path
import re

# ===== 配置 =====
API_KEY = os.environ.get("API_KEY", "")
BASE_URL = os.environ.get("BASE_URL", "https://api.viviai.cc/v1")

# 精简版规则文件
RULES_FILE = Path(__file__).parent.parent / "合体字规则_精简版.md"

rules_cache = None


def load_rules():
    global rules_cache
    if rules_cache is not None:
        return rules_cache
    try:
        with open(RULES_FILE, "r", encoding="utf-8") as f:
            rules_cache = f.read()
    except FileNotFoundError:
        print(f"警告: 规则文件不存在 {RULES_FILE}")
        rules_cache = ""
    return rules_cache


def build_prompt(characters):
    rules = load_rules()
    word = "".join(characters)

    return f"""{rules}

现在请将以下4个字合成为一个合体字：{word}

要求：
1. 四个字的部件必须拆散重组，通过共享笔画融合为一个整体，不是简单地把四个完整的字摆在一起
2. 整体构图为正方形
3. 风格为传统毛笔书法，有墨迹质感和笔锋变化
4. 正面平视角度，90度正交俯视，无透视，无3D效果
5. 请直接生成图片"""


def call_api(prompt, attempt_number):
    variant = "（版本A：注重传统结构融合）" if attempt_number == 1 else "（版本B：注重创意意象表达）"
    final_prompt = prompt + variant

    payload = json.dumps({
        "model": "gpt-image-2",
        "messages": [{"role": "user", "content": final_prompt}],
        "max_tokens": 4096,
        "temperature": 0.8,
        "top_p": 1,
        "frequency_penalty": 0.1
    }).encode("utf-8")

    req = Request(
        f"{BASE_URL}/chat/completions",
        data=payload,
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {API_KEY}"
        },
        method="POST"
    )

    try:
        with urlopen(req, timeout=180) as resp:
            data = json.loads(resp.read().decode("utf-8"))

        if "choices" in data and len(data["choices"]) > 0:
            message = data["choices"][0].get("message", {})
            content = message.get("content", "")

            # 检查 parts 中的 inline 图片
            for part in message.get("parts", []):
                if isinstance(part, dict):
                    inline_data = part.get("inline_data") or part.get("inlineData")
                    if inline_data:
                        mime = inline_data.get("mime_type", inline_data.get("mimeType", "image/png"))
                        b64 = inline_data.get("data", "")
                        if b64:
                            return f"data:{mime};base64,{b64}"

            # 匹配 content 中的 base64 图片
            if content:
                b64_match = re.search(r'data:image/[^;]+;base64,([A-Za-z0-9+/=]+)', content)
                if b64_match:
                    return b64_match.group(0)

                # markdown 格式 ![image](data:image/...)
                md_match = re.search(r'!\[.*?\]\((data:image/[^\)]+)\)', content)
                if md_match:
                    return md_match.group(1)

                url_match = re.search(r'(https?://[^\s\)]+\.(?:png|jpg|jpeg|webp))', content)
                if url_match:
                    return url_match.group(1)

        raise Exception("响应中未找到图片")

    except HTTPError as e:
        error_body = e.read().decode("utf-8", errors="replace")
        raise Exception(f"API错误 {e.code}: {error_body[:200]}")
    except URLError as e:
        raise Exception(f"网络错误: {e.reason}")


def generate_images(characters):
    prompt = build_prompt(characters)
    image = call_api(prompt, 1)
    return [image]


class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length)
            data = json.loads(body.decode("utf-8"))

            characters = data.get("characters", [])

            if not characters or len(characters) != 4:
                self._send({"success": False, "error": "请提供4个汉字"}, 400)
                return

            for ch in characters:
                if not ch or len(ch) != 1 or ord(ch) < 0x4E00 or ord(ch) > 0x9FFF:
                    self._send({"success": False, "error": f"'{ch}' 不是有效的汉字"}, 400)
                    return

            images = generate_images(characters)
            self._send({"success": True, "images": images})

        except json.JSONDecodeError:
            self._send({"success": False, "error": "请求格式错误"}, 400)
        except Exception as e:
            self._send({"success": False, "error": str(e)}, 500)

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.end_headers()

    def _send(self, data, status=200):
        body = json.dumps(data, ensure_ascii=False).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(body)

File: backend/generate.js
const fs = require('fs').promises;
const path = require('path');

// 规则文件路径
const RULES_DIR = path.join(__dirname, 'rules');
const RULE_FILES = [
  '拓扑编码库.md',
  '几何优化库.md',
  '风格特征库.md',
  '语义校验库.md'
];

// 缓存规则内容
let rulesCache = null;

/**
 * 加载所有规则文档
 */
async function loadRules() {
  if (rulesCache) return rulesCache;

  const rules = {};

  for (const file of RULE_FILES) {
    try {
      const content = await fs.readFile(path.join(RULES_DIR, file), 'utf-8');
      rules[file.replace('.md', '')] = content;
    } catch (err) {
      console.warn(`警告: 无法读取规则文件 ${file}`, err.message);
      rules[file.replace('.md', '')] = '';
    }
  }

  rulesCache = rules;
  return rules;
}

/**
 * 构建Prompt
 */
async function buildPrompt(characters) {
  const rules = await loadRules();
  const word = characters.join('');

  return `你是一个专业的中国书法合体字设计师。你需要根据以下四个规则库，将用户给你的4个汉字合成为一个艺术合体字。

== 拓扑编码库 ==
${rules['拓扑编码库']}

== 几何优化库 ==
${rules['几何优化库']}

== 风格特征库 ==
${rules['风格特征库']}

== 语义校验库 ==
${rules['语义校验库']}

现在请将以下4个字合成为一个合体字：${word}

要求：
1. 四个字的部件必须拆散重组，通过共享笔画融合为一个整体，不是简单地把四个完整的字摆在一起
2. 整体构图为正方形
3. 风格为传统毛笔书法，有墨迹质感和笔锋变化
4. 正面平视角度，90度正交俯视，无透视，无3D效果
5. 可以根据词语含义进行意象化设计（如将特征性的笔画异化为与词义相关的图形）
6. 请直接生成图片`;
}

/**
 * 通过 OpenAI 兼容 API 生成图片
 */
async function generateViaOpenAI(prompt) {
  const apiKey = process.env.API_KEY;
  const baseURL = process.env.BASE_URL || 'https://api.viviai.cc/v1';

  if (!apiKey) {
    throw new Error('未设置 API_KEY');
  }

  const response = await fetch(`${baseURL}/images/generations`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${apiKey}`
    },
    body: JSON.stringify({
      model: 'nanobanana-2',  // 或者根据实际支持的模型名称调整
      prompt: prompt,
      n: 1,
      size: '1024x1024',
      response_format: 'b64_json'
    })
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`API 请求失败 (${response.status}): ${errorText}`);
  }

  const data = await response.json();

  // 处理不同格式的响应
  if (data.data && data.data[0]) {
    if (data.data[0].b64_json) {
      return `data:image/png;base64,${data.data[0].b64_json}`;
    }
    if (data.data[0].url) {
      return data.data[0].url;
    }
  }

  throw new Error('无法从响应中提取图片');
}

/**
 * 生成单张图片
 */
async function generateSingleImage(characters, attemptNumber) {
  const prompt = await buildPrompt(characters);

  // 添加随机性，确保两次生成不同
  const variantHint = attemptNumber === 1
    ? '（版本A：注重传统结构融合）'
    : '（版本B：注重创意意象表达）';

  const finalPrompt = prompt + variantHint;

  console.log(`  [尝试 ${attemptNumber}] 调用 API...`);
  console.log(`  [尝试 ${attemptNumber}] 提示词长度: ${finalPrompt.length} 字符`);

  try {
    const imageData = await generateViaOpenAI(finalPrompt);
    console.log(`  [尝试 ${attemptNumber}] 图片生成成功`);
    return imageData;
  } catch (error) {
    console.error(`  [尝试 ${attemptNumber}] 失败:`, error.message);
    throw error;
  }
}

/**
 * 并行生成2张图片
 */
async function generateImages(characters) {
  const word = characters.join('');
  console.log(`开始为 "${word}" 生成2张图片...`);

  // 并行执行两次生成
  const promises = [
    generateSingleImage(characters, 1),
    generateSingleImage(characters, 2)
  ];

  const results = await Promise.allSettled(promises);

  const images = [];
  const errors = [];

  results.forEach((result, index) => {
    if (result.status === 'fulfilled') {
      images.push(result.value);
    } else {
      errors.push(`图片${index + 1}: ${result.reason.message}`);
    }
  });

  // 如果至少有一张成功，返回结果
  if (images.length > 0) {
    console.log(`成功生成 ${images.length} 张图片`);
    return images;
  }

  // 全部失败时抛出错误
  throw new Error('图片生成失败: ' + errors.join('; '));
}

module.exports = {
  generateImages
};
